File: 2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.java
class Solution {
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int n = A.length, cnt = 0;
        int[] res = new int[n], freq = new int[n+1];
        for (int i=0; i<n; i++) {
            freq[A[i]]++;
            if (freq[A[i]] == 2)    cnt++;
            freq[B[i]]++;
            if (freq[B[i]] == 2)    cnt++;
            res[i] = cnt;
        }
        return res;
    }
}

File: 2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.py
class Solution:
    def findThePrefixCommonArray(self, A: List[int], B: List[int]) -> List[int]:
        seta, setb = set(), set()
        cnt, res, i = 0, [0] * len(A), 0
        for a,b in zip(A,B):
            seta.add(a)
            setb.add(b)
            if a != b and a in setb and b in seta:
                cnt += 2
            elif a == b:
                cnt += 1
            elif a in setb or b in seta:
                cnt += 1
            res[i] = cnt
            i += 1
        return res